### PR TITLE
3D support

### PR DESCRIPTION
**Pull request by Julien-Sahli**
Re-upload of https://github.com/LCAV/LenslessPiCam/tree/3D

## 3D support for Gradient Descent (and APGD soon), also with pytorch

Benchmarks done with a AMD Ryzen 7 5800H CPU and a NVIDIA Geforce RTX 3050 Laptop GPU

Data used : see `data.rst`

### Gradient Descent, 2D

Default : **~140 s**
python scripts/recon/gradient_descent.py

Torch, CPU :  **~70 s**
python scripts/recon/gradient_descent.py torch=True

Torch, cuda:0 : **~22 s**
python scripts/recon/gradient_descent.py -cn pytorch

### Gradient Descent, 3D 

Default : **~135 s**
python scripts/recon/gradient_descent.py input.psf="data/psf/diffuser_cam.npy" input.data="data/raw_data/diffuser_cam.tiff"  preprocess.downsample=1

Torch, CPU : **~105 s**
python scripts/recon/gradient_descent.py torch=True input.psf="data/psf/diffuser_cam.npy" input.data="data/raw_data/diffuser_cam.tiff"  preprocess.downsample=1

Torch, cuda:0 : **~27 s**
python scripts/recon/gradient_descent.py -cn pytorch input.psf="data/psf/diffuser_cam.npy" input.data="data/raw_data/diffuser_cam.tiff"  preprocess.downsample=1

### CODE REVIEW

File: lensless/admm.py [hunk 1]
```diff
@@ -191,11 +199,11 @@ def _image_update(self):
         # rk = self._convolver._pad(rk)
 
         if self.is_torch:
-            freq_space_result = self._R_divmat * torch.fft.rfft2(rk, dim=(0, 1))
-            self._image_est = torch.fft.irfft2(freq_space_result, dim=(0, 1))
+            freq_space_result = self._R_divmat * torch.fft.rfft2(rk, dim=(1, 2))
```
Comment (ebezzam) [inline lensless/admm.py:202]: change to negative indices

File: lensless/admm.py [hunk 2]
```diff
@@ -288,10 +295,16 @@ def finite_diff_gram(shape, dtype=None, is_torch=False):
...
+        gram[0, 0, 1] = gram[0, 0, -1] = gram[0, 1, 0] = gram[0, -1, 0] = -1
+    else:
+        gram[0, 0, 0] = 6
+        gram[0, 0, 1] = gram[0, 0, -1] = gram[0, 1, 0] = gram[0, -1, 0] = gram[1, 0, 0] = gram[
+            -1, 0, 0
+        ] = -1
 
     if is_torch:
-        return torch.fft.rfft2(gram, dim=(0, 1))
+        return torch.fft.rfft2(gram, dim=(1, 2))
```
Comment (ebezzam) [inline lensless/admm.py:308]: negative indices

File: lensless/gradient_descent.py [hunk 3]
```diff
@@ -50,7 +50,7 @@ def non_neg(xi):
 
 
 class GradientDescent(ReconstructionAlgorithm):
-    def __init__(self, psf, dtype=None, proj=non_neg, **kwargs):
+    def __init__(self, psf, dtype=None, proj=non_neg, initial_est=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (ebezzam) [reply lensless/gradient_descent.py:53]: I'm leaning towards not giving this option, unless you saw a case where it was useful?

File: lensless/io.py [hunk 4]
```diff
@@ -163,57 +164,88 @@ def load_psf(
...
     # resize
-    if shape:
-        psf = resize(psf, shape=shape)
-    elif downsample != 1:
-        psf = resize(psf, factor=1 / downsample)
+    psf = resize(psf, shape=shape, factor=1 / downsample)
 
     if single_psf:
-        if len(psf.shape) == 3:
+        if psf.shape[3] > 1:
```
Comment (ebezzam) [inline lensless/io.py:244]: `if not grayscale`?

File: lensless/recon.py [hunk 5]
```diff
@@ -114,6 +114,27 @@ class :py:class:`~lensless.ReconstructionAlgorithm`. The three reconstruction
     python scripts/recon/admm.py -cn <CONFIG_FILENAME_WITHOUT_YAML_EXT>
 
 
+3D example
+----------
+
+It is also possible to reconstruct 3D scenes using :py:class:`~lensless.GradientDescent` or :py:class:`~lensless.APGD`. :py:class:`~lensless.ADMM` doesn't supports 3D reconstruction yet.
```
Comment (ebezzam) [inline lensless/recon.py:120]: "does not support"

File: lensless/recon.py [hunk 6]
```diff
@@ -114,6 +114,27 @@ class :py:class:`~lensless.ReconstructionAlgorithm`. The three reconstruction
...
+This requires to use a 3D PSF as an input in the form of an .npy or .npz file, which actually is a set of 2D PSFs corresponding to the same diffuser sampled with light sources from different depths.
+The input data for 3D reconstructions is still a 2D image, as collected by the camera. The reconstruction will be able to separate which part of the lensless data corresponds to which 2D PSF,
+and therefore to which depth, effectively generating a 3D reconstruction, which will be outputed in the form of an .npy file. A 2D projection on the depth axis is also displayed to the user.
+
+The same scripts for 2D reconstruction can be used for 3D reconstruction, namely ``scripts/recon/gradient_descent.py`` and ``scripts/recon/apgd_pycsou.py``.
+
+3D data is provided in LenslessPiCam, but it is simulated. Real example data can be obtained from `Waller Lab <https://github.com/Waller-Lab/DiffuserCam/tree/master/example_data>`_.
+For both the simulated data and the data from Waller Lab, it is best to set ``downsample=1`` :
```
Comment (ebezzam) [inline lensless/recon.py:128]: remove space before ":"

File: docs/source/data.rst [hunk 7]
```diff
@@ -54,3 +59,33 @@ dataset (200 files, 725 MB). It was prepared with the following script:
 
     python scripts/prepare_mirflickr_subset.py \
     --data ~/Documents/DiffuserCam/DiffuserCam_Mirflickr_Dataset
+
+
+You can download example 3D PSF and raw data from the Waller lab
+`here  <https://github.com/Waller-Lab/DiffuserCam/tree/master/example_data>`__.
+The PSF has to be converted from .mat to .npy in order to be usable :
```
Comment (ebezzam) [inline docs/source/data.rst:66]: remove space before ":"

File: docs/source/data.rst [hunk 8]
```diff
@@ -54,3 +59,33 @@ dataset (200 files, 725 MB). It was prepared with the following script:
...
+.. code:: bash
+
+	python scripts/data/3d/npy_to_tiff.py ~path/to/output.npy
+
+
+You may also want to export it into a wavefront .obj file
+for it to be displayed in 3D rendering softwares with the following
+script. It mostly exists to allow the user to preview it and is not
+totally accurate as the problem of converting discrete pixels in a
+"continuous" wavefront object is subject to interpretation :
```
Comment (ebezzam) [inline docs/source/data.rst:86]: remove space before ":"

File: lensless/recon.py [hunk 9]
```diff
@@ -291,14 +305,11 @@ def set_data(self, data):
...
-        if not self._is_rgb:
-            assert len(data.shape) == 2
-            data = data[:, :, None]
-        assert len(self._psf_shape) == len(data.shape)
+        assert len(data.shape) == 4, "Data must be 4D: [depth, width, height, channel]."
 
         # assert same shapes
         assert np.all(
-            self._psf_shape[:2] == np.array(data.shape)[:2]
+            self._psf_shape[1:3] == np.array(data.shape)[1:3]
```
Comment (ebezzam) [inline lensless/recon.py:312]: negative indices

File: lensless/rfft_convolve.py [hunk 10]
```diff
@@ -32,15 +32,11 @@ def __init__(self, psf, dtype=None, pad=True, norm="ortho", **kwargs):
...
-            if psf.shape[2] == 3:
-                self._is_rgb = True
-            self._psf = psf
-        else:
-            self._psf = psf[:, :, None]
-        self._n_channels = self._psf.shape[2]
-        self._psf_shape = np.array(self._psf.shape)
+
+        assert len(psf.shape) == 4, "Expected 4D PSF of shape (depth, width, height, channels)"
+        self._use_3d = psf.shape[0] != 1
```
Comment (ebezzam) [inline lensless/rfft_convolve.py:37]: negative indices

File: lensless/util.py [hunk 11]
```diff
@@ -28,18 +35,36 @@ def resize(img, factor=None, shape=None, interpolation=cv2.INTER_CUBIC):
...
+    assert len(img.shape) == 4, "Image must be of shape (depth, height, width, color)"
     min_val = img.min()
     max_val = img.max()
-    img_shape = np.array(img.shape)[:2]
-    if shape is None:
-        assert factor is not None
-        new_shape = tuple((img_shape * factor).astype(int))
-        new_shape = new_shape[::-1]
-        resized = cv2.resize(img, dsize=new_shape, interpolation=interpolation)
+    img_shape = np.array(img.shape)[1:3]
```
Comment (ebezzam) [inline lensless/util.py:41]: negative indices

File: lensless/util.py [hunk 12]
```diff
@@ -28,18 +35,36 @@ def resize(img, factor=None, shape=None, interpolation=cv2.INTER_CUBIC):
...
+
+    assert not ((factor is None) and (shape is None)), "Must specify either factor or shape"
+    new_shape = tuple((img_shape * factor).astype(int)) if (shape is None) else shape[1:3]
+
+    if np.array_equal(img_shape, new_shape):
+        return img
+
+    if torch_available:
+        # torch resize expects an input of form [color, depth, width, height]
+        tmp = np.moveaxis(img, 3, 0)
```
Comment (ebezzam) [inline lensless/util.py:51]: negative indices

File: lensless/util.py [hunk 13]
```diff
@@ -28,18 +35,36 @@ def resize(img, factor=None, shape=None, interpolation=cv2.INTER_CUBIC):
...
+        resized = np.moveaxis(resized, 0, 3)
+
     else:
-        if np.array_equal(img_shape, shape[::-1]):
-            return img
-        resized = cv2.resize(img, dsize=shape[::-1], interpolation=interpolation)
+        resized = np.array(
+            [
+                cv2.resize(img[i], dsize=new_shape[::-1], interpolation=interpolation)
+                for i in range(img.shape[0])
```
Comment (ebezzam) [inline lensless/util.py:61]: negative index

File: lensless/util.py [hunk 14]
```diff
@@ -28,18 +35,36 @@ def resize(img, factor=None, shape=None, interpolation=cv2.INTER_CUBIC):
...
+            ]
+        )
+        # OpenCV discards channel dimension if it is 1, put it back
+        if len(resized.shape) == 3:
+            resized = resized[:, :, :, np.newaxis]
```
Comment (ebezzam) [inline lensless/util.py:66]: might be best to use [expand dims](https://numpy.org/doc/stable/reference/generated/numpy.expand_dims.html) like this: 

```
resized = np.expand_dims(resized, axis=-1)
```

File: test/test_algos.py [hunk 15]
```diff
@@ -43,10 +43,7 @@ def test_gradient_descent():
                 recon = GradientDescent(psf, dtype=dtype)
                 recon.set_data(data)
                 res = recon.apply(n_iter=n_iter, disp_iter=None, plot=False)
-                if gray:
-                    assert len(psf.shape) == 2
-                else:
-                    assert len(psf.shape) == 3
+                assert psf.shape[3] == (1 if gray else 3)
```
Comment (ebezzam) [inline test/test_algos.py:46]: negative index
